File: helfomat-api/src/main/java/de/helfenkannjeder/helfomat/api/organisation/event/OrganizationEventDtoAssembler.java
package de.helfenkannjeder.helfomat.api.organisation.event;

import de.helfenkannjeder.helfomat.api.organisation.OrganisationAssembler;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationCreateEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationDeleteEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddAddressEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddAttendanceTimeEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddContactPersonEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddGroupEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddPictureEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddQuestionAnswerEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditAddVolunteerEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDefaultAddressEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDeleteAttendanceTimeEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDeleteContactPersonEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDeleteGroupEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDeletePictureEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDeleteQuestionAnswerEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDeleteVolunteerEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditDescriptionEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditLogoEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditNameEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditTeaserImageEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditUrlNameEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEditWebsiteEvent;
import de.helfenkannjeder.helfomat.core.organisation.event.OrganisationEvent;

import java.util.List;
import java.util.stream.Collectors;

/**
 * @author Valentin Zickner
 */
public class OrganizationEventDtoAssembler implements OrganizationEventDtoVisitor<OrganisationEvent> {

    public static List<OrganisationEvent> toOrganizationEvent(List<OrganizationEventDto> organizationEvents) {
        OrganizationEventDtoAssembler organizationEventDtoAssembler = new OrganizationEventDtoAssembler();
        return organizationEvents.stream()
            .map(organisationEvent -> organisationEvent.visit(organizationEventDtoAssembler))
            .collect(Collectors.toList());
    }

    @Override
    public OrganisationEvent visit(OrganizationCreateEventDto organizationCreateEventDto) {
        return new OrganisationCreateEvent(
            organizationCreateEventDto.getOrganisationId(),
            organizationCreateEventDto.getName(),
            organizationCreateEventDto.getUrlName(),
            organizationCreateEventDto.getOrganisationType()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationDeleteEventDto organizationDeleteEventDto) {
        return new OrganisationDeleteEvent(
            organizationDeleteEventDto.getOrganisationId()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddAddressEventDto organizationEditAddAddressEventDto) {
        return new OrganisationEditAddAddressEvent(
            organizationEditAddAddressEventDto.getOrganisationId(),
            organizationEditAddAddressEventDto.getIndex(),
            OrganisationAssembler.toAddress(organizationEditAddAddressEventDto.getAddress())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddAttendanceTimeEventDto organizationEditAddAttendanceTimeEventDto) {
        return new OrganisationEditAddAttendanceTimeEvent(
            organizationEditAddAttendanceTimeEventDto.getOrganisationId(),
            organizationEditAddAttendanceTimeEventDto.getIndex(),
            OrganisationAssembler.toAttendanceTime(organizationEditAddAttendanceTimeEventDto.getAttendanceTime())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddContactPersonEventDto organizationEditAddContactPersonEventDto) {
        return new OrganisationEditAddContactPersonEvent(
            organizationEditAddContactPersonEventDto.getOrganisationId(),
            organizationEditAddContactPersonEventDto.getIndex(),
            OrganisationAssembler.toContactPerson(organizationEditAddContactPersonEventDto.getContactPerson())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddGroupEventDto organizationEditAddGroupEventDto) {
        return new OrganisationEditAddGroupEvent(
            organizationEditAddGroupEventDto.getOrganisationId(),
            organizationEditAddGroupEventDto.getIndex(),
            OrganisationAssembler.toGroup(organizationEditAddGroupEventDto.getGroup())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddPictureEventDto organizationEditAddPictureEventDto) {
        return new OrganisationEditAddPictureEvent(
            organizationEditAddPictureEventDto.getOrganisationId(),
            organizationEditAddPictureEventDto.getIndex(),
            organizationEditAddPictureEventDto.getPictureId()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddQuestionAnswerEventDto organizationEditAddQuestionAnswerEventDto) {
        return new OrganisationEditAddQuestionAnswerEvent(
            organizationEditAddQuestionAnswerEventDto.getOrganisationId(),
            organizationEditAddQuestionAnswerEventDto.getIndex(),
            OrganisationAssembler.toQuestionAnswer(organizationEditAddQuestionAnswerEventDto.getAnsweredQuestion())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditAddVolunteerEventDto organizationEditAddVolunteerEventDto) {
        return new OrganisationEditAddVolunteerEvent(
            organizationEditAddVolunteerEventDto.getOrganisationId(),
            organizationEditAddVolunteerEventDto.getIndex(),
            OrganisationAssembler.toVolunteer(organizationEditAddVolunteerEventDto.getVolunteer())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDefaultAddressEventDto organizationEditDefaultAddressEventDto) {
        return new OrganisationEditDefaultAddressEvent(
            organizationEditDefaultAddressEventDto.getOrganisationId(),
            OrganisationAssembler.toAddress(organizationEditDefaultAddressEventDto.getAddress())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeleteAddressEventDto organizationEditDeleteAddressEventDto) {
        return new OrganisationEditDefaultAddressEvent(
            organizationEditDeleteAddressEventDto.getOrganisationId(),
            OrganisationAssembler.toAddress(organizationEditDeleteAddressEventDto.getAddress())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeleteAttendanceTimeEventDto organizationEditDeleteAttendanceTimeEventDto) {
        return new OrganisationEditDeleteAttendanceTimeEvent(
            organizationEditDeleteAttendanceTimeEventDto.getOrganisationId(),
            OrganisationAssembler.toAttendanceTime(organizationEditDeleteAttendanceTimeEventDto.getAttendanceTime())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeleteContactPersonEventDto organizationEditDeleteContactPersonEventDto) {
        return new OrganisationEditDeleteContactPersonEvent(
            organizationEditDeleteContactPersonEventDto.getOrganisationId(),
            OrganisationAssembler.toContactPerson(organizationEditDeleteContactPersonEventDto.getContactPerson())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeleteGroupEventDto organizationEditDeleteGroupEventDto) {
        return new OrganisationEditDeleteGroupEvent(
            organizationEditDeleteGroupEventDto.getOrganisationId(),
            OrganisationAssembler.toGroup(organizationEditDeleteGroupEventDto.getGroup())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeletePictureEventDto organizationEditDeletePictureEventDto) {
        return new OrganisationEditDeletePictureEvent(
            organizationEditDeletePictureEventDto.getOrganisationId(),
            organizationEditDeletePictureEventDto.getPictureId()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeleteQuestionAnswerEventDto organizationEditDeleteQuestionAnswerEventDto) {
        return new OrganisationEditDeleteQuestionAnswerEvent(
            organizationEditDeleteQuestionAnswerEventDto.getOrganisationId(),
            OrganisationAssembler.toQuestionAnswer(organizationEditDeleteQuestionAnswerEventDto.getAnsweredQuestion())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDeleteVolunteerEventDto organizationEditDeleteVolunteerEventDto) {
        return new OrganisationEditDeleteVolunteerEvent(
            organizationEditDeleteVolunteerEventDto.getOrganisationId(),
            OrganisationAssembler.toVolunteer(organizationEditDeleteVolunteerEventDto.getVolunteer())
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditDescriptionEventDto organizationEditDescriptionEventDto) {
        return new OrganisationEditDescriptionEvent(
            organizationEditDescriptionEventDto.getOrganisationId(),
            organizationEditDescriptionEventDto.getDescription()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditLogoEventDto organizationEditLogoEventDto) {
        return new OrganisationEditLogoEvent(
            organizationEditLogoEventDto.getOrganisationId(),
            organizationEditLogoEventDto.getLogo()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditNameEventDto organizationEditNameEventDto) {
        return new OrganisationEditNameEvent(
            organizationEditNameEventDto.getOrganisationId(),
            organizationEditNameEventDto.getName()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditTeaserImageEventDto organizationEditTeaserImageEventDto) {
        return new OrganisationEditTeaserImageEvent(
            organizationEditTeaserImageEventDto.getOrganisationId(),
            organizationEditTeaserImageEventDto.getTeaserImage()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditUrlNameEventDto organizationEditUrlNameEventDto) {
        return new OrganisationEditUrlNameEvent(
            organizationEditUrlNameEventDto.getOrganisationId(),
            organizationEditUrlNameEventDto.getUrlName()
        );
    }

    @Override
    public OrganisationEvent visit(OrganizationEditWebsiteEventDto organizationEditWebsiteEventDto) {
        return new OrganisationEditWebsiteEvent(
            organizationEditWebsiteEventDto.getOrganisationId(),
            organizationEditWebsiteEventDto.getWebsite()
        );
    }
}

File: helfomat-api/src/main/kotlin/de/helfenkannjeder/helfomat/api/organization/event/OrganizationEventDtoAssembler.kt
package de.helfenkannjeder.helfomat.api.organization.event

import de.helfenkannjeder.helfomat.api.organization.*
import de.helfenkannjeder.helfomat.core.organization.event.*

/**
 * @author Valentin Zickner
 */
class OrganizationEventDtoAssembler : OrganizationEventDtoVisitor<OrganizationEvent> {

    override fun visit(organizationCreateEventDto: OrganizationCreateEventDto) =
        OrganizationCreateEvent(organizationCreateEventDto.organizationId, organizationCreateEventDto.name, organizationCreateEventDto.urlName, organizationCreateEventDto.organizationType)

    override fun visit(organizationDeleteEventDto: OrganizationDeleteEventDto) =
        OrganizationDeleteEvent(organizationDeleteEventDto.organizationId)

    override fun visit(organizationEditAddAddressEventDto: OrganizationEditAddAddressEventDto) =
        OrganizationEditAddAddressEvent(organizationEditAddAddressEventDto.organizationId, organizationEditAddAddressEventDto.index, organizationEditAddAddressEventDto.address.toAddress())

    override fun visit(organizationEditAddAttendanceTimeEventDto: OrganizationEditAddAttendanceTimeEventDto) =
        OrganizationEditAddAttendanceTimeEvent(organizationEditAddAttendanceTimeEventDto.organizationId, organizationEditAddAttendanceTimeEventDto.index, organizationEditAddAttendanceTimeEventDto.attendanceTime.toAttendanceTime())

    override fun visit(organizationEditAddContactPersonEventDto: OrganizationEditAddContactPersonEventDto) =
        OrganizationEditAddContactPersonEvent(organizationEditAddContactPersonEventDto.organizationId, organizationEditAddContactPersonEventDto.index, organizationEditAddContactPersonEventDto.contactPerson.toContactPerson())

    override fun visit(organizationEditAddGroupEventDto: OrganizationEditAddGroupEventDto) =
        OrganizationEditAddGroupEvent(organizationEditAddGroupEventDto.organizationId, organizationEditAddGroupEventDto.index, organizationEditAddGroupEventDto.group.toGroup())

    override fun visit(organizationEditAddPictureEventDto: OrganizationEditAddPictureEventDto) =
        OrganizationEditAddPictureEvent(organizationEditAddPictureEventDto.organizationId, organizationEditAddPictureEventDto.index, organizationEditAddPictureEventDto.pictureId)

    override fun visit(organizationEditAddQuestionAnswerEventDto: OrganizationEditAddQuestionAnswerEventDto) =
        OrganizationEditAddQuestionAnswerEvent(organizationEditAddQuestionAnswerEventDto.organizationId, organizationEditAddQuestionAnswerEventDto.index, organizationEditAddQuestionAnswerEventDto.answeredQuestion.toQuestionAnswer())

    override fun visit(organizationEditAddVolunteerEventDto: OrganizationEditAddVolunteerEventDto) =
        OrganizationEditAddVolunteerEvent(organizationEditAddVolunteerEventDto.organizationId, organizationEditAddVolunteerEventDto.index, organizationEditAddVolunteerEventDto.volunteer.toVolunteer())

    override fun visit(organizationEditDefaultAddressEventDto: OrganizationEditDefaultAddressEventDto) =
        OrganizationEditDefaultAddressEvent(organizationEditDefaultAddressEventDto.organizationId, organizationEditDefaultAddressEventDto.address?.toAddress())

    override fun visit(organizationEditDeleteAddressEventDto: OrganizationEditDeleteAddressEventDto) =
        OrganizationEditDefaultAddressEvent(organizationEditDeleteAddressEventDto.organizationId, organizationEditDeleteAddressEventDto.address.toAddress())

    override fun visit(organizationEditDeleteAttendanceTimeEventDto: OrganizationEditDeleteAttendanceTimeEventDto) =
        OrganizationEditDeleteAttendanceTimeEvent(organizationEditDeleteAttendanceTimeEventDto.organizationId, organizationEditDeleteAttendanceTimeEventDto.attendanceTime.toAttendanceTime())

    override fun visit(organizationEditDeleteContactPersonEventDto: OrganizationEditDeleteContactPersonEventDto) =
        OrganizationEditDeleteContactPersonEvent(organizationEditDeleteContactPersonEventDto.organizationId, organizationEditDeleteContactPersonEventDto.contactPerson.toContactPerson())

    override fun visit(organizationEditDeleteGroupEventDto: OrganizationEditDeleteGroupEventDto) =
        OrganizationEditDeleteGroupEvent(organizationEditDeleteGroupEventDto.organizationId, organizationEditDeleteGroupEventDto.group.toGroup())

    override fun visit(organizationEditDeletePictureEventDto: OrganizationEditDeletePictureEventDto) =
        OrganizationEditDeletePictureEvent(organizationEditDeletePictureEventDto.organizationId, organizationEditDeletePictureEventDto.pictureId)

    override fun visit(organizationEditDeleteQuestionAnswerEventDto: OrganizationEditDeleteQuestionAnswerEventDto) =
        OrganizationEditDeleteQuestionAnswerEvent(organizationEditDeleteQuestionAnswerEventDto.organizationId, organizationEditDeleteQuestionAnswerEventDto.answeredQuestion.toQuestionAnswer())

    override fun visit(organizationEditDeleteVolunteerEventDto: OrganizationEditDeleteVolunteerEventDto) =
        OrganizationEditDeleteVolunteerEvent(organizationEditDeleteVolunteerEventDto.organizationId, organizationEditDeleteVolunteerEventDto.volunteer.toVolunteer())

    override fun visit(organizationEditDescriptionEventDto: OrganizationEditDescriptionEventDto) =
        OrganizationEditDescriptionEvent(organizationEditDescriptionEventDto.organizationId, organizationEditDescriptionEventDto.description)

    override fun visit(organizationEditLogoEventDto: OrganizationEditLogoEventDto) =
        OrganizationEditLogoEvent(organizationEditLogoEventDto.organizationId, organizationEditLogoEventDto.logo)

    override fun visit(organizationEditNameEventDto: OrganizationEditNameEventDto) =
        OrganizationEditNameEvent(organizationEditNameEventDto.organizationId, organizationEditNameEventDto.name)

    override fun visit(organizationEditTeaserImageEventDto: OrganizationEditTeaserImageEventDto) =
        OrganizationEditTeaserImageEvent(organizationEditTeaserImageEventDto.organizationId, organizationEditTeaserImageEventDto.teaserImage)

    override fun visit(organizationEditUrlNameEventDto: OrganizationEditUrlNameEventDto) =
        OrganizationEditUrlNameEvent(organizationEditUrlNameEventDto.organizationId, organizationEditUrlNameEventDto.urlName)

    override fun visit(organizationEditWebsiteEventDto: OrganizationEditWebsiteEventDto) =
        OrganizationEditWebsiteEvent(organizationEditWebsiteEventDto.organizationId, organizationEditWebsiteEventDto.website)

    override fun visit(organizationEditChangeGroupEventDto: OrganizationEditChangeGroupEventDto): OrganizationEvent =
        OrganizationEditChangeGroupEvent(organizationEditChangeGroupEventDto.organizationId, organizationEditChangeGroupEventDto.indexOffset, organizationEditChangeGroupEventDto.oldGroup.toGroup(), organizationEditChangeGroupEventDto.group.toGroup())

    override fun visit(organizationEditChangePictureEventDto: OrganizationEditChangePictureEventDto): OrganizationEvent =
        OrganizationEditChangePictureEvent(organizationEditChangePictureEventDto.organizationId, organizationEditChangePictureEventDto.indexOffset, organizationEditChangePictureEventDto.pictureId)

    override fun visit(organizationEditChangeAttendanceTimeEventDto: OrganizationEditChangeAttendanceTimeEventDto): OrganizationEvent =
        OrganizationEditChangeAttendanceTimeEvent(organizationEditChangeAttendanceTimeEventDto.organizationId, organizationEditChangeAttendanceTimeEventDto.indexOffset, organizationEditChangeAttendanceTimeEventDto.oldAttendanceTime.toAttendanceTime(), organizationEditChangeAttendanceTimeEventDto.attendanceTime.toAttendanceTime())

    override fun visit(organizationEditChangeAddressEventDto: OrganizationEditChangeAddressEventDto): OrganizationEvent =
        OrganizationEditChangeAddressEvent(organizationEditChangeAddressEventDto.organizationId, organizationEditChangeAddressEventDto.indexOffset, organizationEditChangeAddressEventDto.oldAddress.toAddress(), organizationEditChangeAddressEventDto.address.toAddress())

    override fun visit(proposedChangeOrganizationEventDto: ProposedChangeOrganizationEventDto) =
        ProposedChangeOrganizationEvent(
            proposedChangeOrganizationEventDto.organizationId,
            proposedChangeOrganizationEventDto.author,
            proposedChangeOrganizationEventDto.sources,
            proposedChangeOrganizationEventDto.changes.map { it.visit(this) }
        )

    override fun visit(confirmedChangeOrganizationEventDto: ConfirmedChangeOrganizationEventDto) =
        ConfirmedChangeOrganizationEvent(
            confirmedChangeOrganizationEventDto.organizationId,
            confirmedChangeOrganizationEventDto.approvedBy,
            confirmedChangeOrganizationEventDto.author,
            confirmedChangeOrganizationEventDto.sources,
            confirmedChangeOrganizationEventDto.changes.map { it.visit(this) }
        )

    companion object {
        fun toOrganizationEvent(organizationEvents: List<OrganizationEventDto>): List<OrganizationEvent> {
            val organizationEventDtoAssembler = OrganizationEventDtoAssembler()
            return organizationEvents.map { it.visit(organizationEventDtoAssembler) }
        }
    }
}